### ISSUE TITLE
Issues with `break`; lexical non-local try/catch label desired

### ISSUE THREAD

**Issue by nicowilliams** [stedolan/jq#658]
The `try/catch` addition is great, and allows for a trivial implementation of `break` for `foreach` and so on.  It's a non-local exit construct, and it's _dynamic_ in nature, which, in a very lexical language, is out of place.  It's one thing to catch bonafide errors, and another to use this mechanism for what should be lexical non-local exits.

Unfortunately, if one tries to `break` from a context where there are unseen `foreach`/... contexts on the stack, then the loop broken won't be the one intended.  A `breakN` won't work either, for the same reason.

About the only thing that will work is a new syntactic construct that creates a unique (see below) "error" to throw to get back to it, and gives it a lexical binding.  Something like:

```
label $<var> | EXP
```

where one might then use like so:

```
label $x | foobar(stuff | if . == "blah" then break $x else . end)
```

Perhaps "label" is a bad term for this...  `goto` certainly is, in various ways, a very bad term for the action of unwinding to a label and producing `empty`, but it is highly recognizable.  (It's not unlike a C with GCC extensions computed goto from a local function to its enclosing block: it's a backtracking/unwinding goto.)

`break $x` does feel natural, and denotes the desired effect: make all of `label $x | ...` produce `empty`.

An anonymous form would be:

```
label | EXP
```

and instead of `break $label` one would use `break` to unwind to the nearest anonymous label.

We can and should make it so that label bindings cannot be used for any other purpose than non-local exits.  This should be an error:

```
label $foo | $foo
```

Internally a label would have to be a normal value which internally is used as with `error`, for throwing an exception to a waiting internal try/catch implied by `label $foo | ...`.

Labels might be values internally, but we should hide them as much as possible.

For language purity reasons (e.g., to allow user-defined control structures as jq-coded functions to use `break` in their argument closures), the builtin iterators like `reduce`, `foreach`, `while`, `until`, `repeat`, and so on, should all _not_ introduce a lexical break-to label.  This is because jq-coded functions can't do that as their argument closures are already fully bound by the point where we could try to introduce the new label.  We might make an exception for `reduce` and `foreach` since they are already special because they involve special syntax and already modify the lexical environment of their "argument" closures.

The thing to throw to break to a label would have unspecified and opaque form.  We can make sure that try/catch never catches these "errors" and `error` can't throw them, but the latter could only produce an error about misuse at runtime!  So we might as well document a naming convention for jq-internal errors to prevent collisions with user-defined errors.

Something like `{__jq: ...}` would work as a naming convention for jq internal uses of errors.  While `{__jq: <counter>}` might be used for labels.

To sum up:
- add `label $label_name | EXP` syntax
- add `break $label_name` syntax to unwind to a named label and make it `empty`
- add `label | EXP` syntax for anonymous labels
- add `break` for unwinding to the lexically-nearest anonymous label
- add `break2` for unwinding to the lexically-second-nearest anonymous label
- add `break3` ...

EDIT: Re-wrote to match the current state of play and direction for this work.
